tests/circuit_graph/test_qasm.py

```diff
@@ -0,0 +1,740 @@
+            qreg q[2];
+            creg c[2];
+            ry(1.5707963267948968) q[1];
+            ry(1.5707963267948963) q[0];
+            cx q[1],q[0];
+            ry(0.0) q[0];
+            cx q[1],q[0];
+            rz(0.0) q[0];
+            cx q[1],q[0];
+            rz(3.141592653589793) q[0];
+            cx q[1],q[0];
+            measure q[0] -> c[0];
+            measure q[1] -> c[1];
+            """
+        )
+
+        assert res == expected
+
+    def test_basis_state_initialization_decomposition(self):
+        """Test that the basis state prepration decomposition
+        is correctly applied."""
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
+        def qnode(state=None):
+            qml.BasisState(state, wires=[0, 1, 2, 3])
+            return qml.expval(qml.PauliZ(0))
+
+        # construct the qnode circuit
+        qnode(state=np.array([1, 0, 1, 1]))
+        res = qnode.circuit.to_openqasm()
+
+        expected = dedent(
+            """\
+            OPENQASM 2.0;
+            include "qelib1.inc";
+            qreg q[4];
+            creg c[4];
+            x q[0];
+            x q[2];
+            x q[3];
+            measure q[0] -> c[0];
+            measure q[1] -> c[1];
+            measure q[2] -> c[2];
+            measure q[3] -> c[3];
+            """
+        )
+
+        assert res == expected
+
+    def test_unsupported_gate(self):
+        """Test an exception is raised if an unsupported operation is
+        applied."""
+        U = np.array([[1, 1], [1, -1]]) / np.sqrt(2)
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def qnode():
+            qml.S(wires=0)
+            qml.QubitUnitary(U, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        qnode()
+
+        with pytest.raises(
+            qml.DeviceError, match="Gate QubitUnitary not supported on device QASM serializer"
+        ):
+            qnode.circuit.to_openqasm()
+
+    def test_rotations(self):
+        """Test that observable rotations are correctly applied."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def qnode():
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            return [
+                qml.expval(qml.PauliX(0)),
+                qml.expval(qml.PauliZ(1)),
+                qml.expval(qml.Hadamard(2)),
+            ]
+
+        qnode()
+        res = qnode.circuit.to_openqasm()
+
+        expected = dedent(
+            """\
+            OPENQASM 2.0;
+            include "qelib1.inc";
+            qreg q[3];
+            creg c[3];
+            h q[0];
+            cx q[0],q[1];
+            h q[0];
+            ry(-0.7853981633974483) q[2];
+            measure q[0] -> c[0];
+            measure q[1] -> c[1];
+            measure q[2] -> c[2];
+            """
+        )
+
+        assert res == expected
+
+
+class TestQASMConformanceTests:

```

Comment:
wondering if this class would belong in PennyLane-Qiskit :thinking:

Reply:
I don't think so, since the logic this function is testing is in PennyLane core, but happy to be convinced otherwise.

Reply:
I think you're right!